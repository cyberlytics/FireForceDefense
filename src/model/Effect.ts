export default abstract class Effect {
    abstract readonly id: string;
    abstract readonly name: string;
    abstract readonly description: string;

    abstract applyEffect(): void;
}
