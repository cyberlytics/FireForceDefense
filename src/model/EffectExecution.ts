import type Effect from './Effect';
import type HexPosition from './HexPosition';

export default class EffectExecution {
    private static nextId = 0;

    public readonly id: number;
    public readonly effect: Effect;
    public readonly position: HexPosition;

    constructor(effect: Effect, position: HexPosition) {
        this.effect = effect;
        this.position = position;
        this.id = EffectExecution.nextId++;
    }
}
