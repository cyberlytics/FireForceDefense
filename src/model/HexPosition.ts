/**
 * This class stores hexagonal positions.
 * It uses axial coordinates (q and r) internally, but also exposes the calculated cube coordinates (x, y, z).
 * @see https://www.redblobgames.com/grids/hexagons/
 */
import HexDirection from './HexDirection';
import type AxialCoordinates from './AxialCoordinates';
import type CubeCoordinates from './CubeCoordinates';

export default class HexPosition implements AxialCoordinates, CubeCoordinates {

    public get x() {
        return this.q;
    }
    public get y() {
        return -this.q - this.r;
    }
    public get z() {
        return this.r;
    }

    /**
     * Constructor for hexagonal positions specified by axial coordinates.
     * @param q The (skewed) column index.
     * @param r The row index.
     */
    public constructor(q: number, r: number) {
        this.q = q;
        this.r = r;
    }

    private static readonly directionVectors = {
        [HexDirection.DIRECTION_0]: new HexPosition(+1, 0),
        [HexDirection.DIRECTION_60]: new HexPosition(+1, -1),
        [HexDirection.DIRECTION_120]: new HexPosition(0, -1),
        [HexDirection.DIRECTION_180]: new HexPosition(-1, 0),
        [HexDirection.DIRECTION_240]: new HexPosition(-1, +1),
        [HexDirection.DIRECTION_300]: new HexPosition(0, +1),
    };
    public readonly q: number;
    public readonly r: number;

    public static fromAxial({q, r}: AxialCoordinates) {
        return new HexPosition(q, r);
    }
    public static fromCube({x, y, z}: CubeCoordinates) {
        if (x + y + z !== 0) {
            console.warn('Invalid coordinates supplied! Ignoring y component.');
        }
        return new HexPosition(x, z);
    }

    public static add(p1: HexPosition, p2: HexPosition) {
        return new HexPosition(p1.q + p2.q, p1.r + p2.r);
    }

    public static getDirectionVector(direction: HexDirection) {
        return HexPosition.directionVectors[direction];
    }

    public getNeighborPosition(direction: HexDirection) {
        return HexPosition.add(this, HexPosition.getDirectionVector(direction));
    }

    public toString() {
        return '(' + this.q + ',' + this.r + ')';
    }
}
