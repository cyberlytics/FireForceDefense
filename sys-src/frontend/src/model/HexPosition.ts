/**
 * This class stores hexagonal positions.
 * It uses axial coordinates (q and r) internally, but also exposes the calculated cube coordinates (x, y, z).
 * @see https://www.redblobgames.com/grids/hexagons/
 */
import HexDirection from './HexDirection';
import type AxialCoordinates from './AxialCoordinates';
import type CubeCoordinates from './CubeCoordinates';

export default class HexPosition implements AxialCoordinates, CubeCoordinates {

    public get x() {
        return this.q;
    }
    public get y() {
        return -this.q - this.r;
    }
    public get z() {
        return this.r;
    }

    /**
     * Constructor for hexagonal positions specified by axial coordinates.
     * @param q The (skewed) column index.
     * @param r The row index.
     */
    public constructor(q: number, r: number) {
        this.q = q;
        this.r = r;
    }

    private static readonly directionVectors = {
        [HexDirection.DIRECTION_0]: new HexPosition(+1, 0),
        [HexDirection.DIRECTION_60]: new HexPosition(+1, -1),
        [HexDirection.DIRECTION_120]: new HexPosition(0, -1),
        [HexDirection.DIRECTION_180]: new HexPosition(-1, 0),
        [HexDirection.DIRECTION_240]: new HexPosition(-1, +1),
        [HexDirection.DIRECTION_300]: new HexPosition(0, +1),
    };
    public readonly q: number;
    public readonly r: number;

    public static fromAxial({q, r}: AxialCoordinates) {
        return new HexPosition(q, r);
    }
    public static fromCube({x, y, z}: CubeCoordinates) {
        if (x + y + z !== 0) {
            console.warn('Invalid coordinates supplied! Ignoring y component.');
        }
        return new HexPosition(x, z);
    }

    public static add(p1: HexPosition, p2: HexPosition) {
        return new HexPosition(p1.q + p2.q, p1.r + p2.r);
    }

    public static getDirectionVector(direction: HexDirection) {
        return HexPosition.directionVectors[direction];
    }

    public static getDistance(p1: HexPosition, p2: HexPosition) {
        return Math.max(
            Math.abs(p1.x - p2.x),
            Math.abs(p1.y - p2.y),
            Math.abs(p1.z - p2.z),
        );
    }

    public static getWidth() {
        return Math.sqrt(3);
    }

    public static getIncircleRadiusHorizontalComponent(hexRadius: number = 0, size: number = 1) {
        return (0.5 + 0.75 * hexRadius - (hexRadius % 2) * 0.25) * this.getWidth() * size;
    }

    public static getIncircleRadiusVerticalComponent(hexRadius: number = 0, size: number = 1) {
        const tmp = Math.floor((hexRadius - 1) / 2) * 0.75 + 0.5;
        return (tmp + Math.abs(tmp)) * size;
    }

    public static getIncircleRadius(hexRadius: number = 0, size: number = 1) {
        return Math.sqrt(
            Math.pow(this.getIncircleRadiusHorizontalComponent(hexRadius, size), 2)
            + Math.pow(this.getIncircleRadiusVerticalComponent(hexRadius, size), 2)
        );
    }

    public getCenterX(size: number = 1) {
        return size * (1.732050808 * this.q + 0.8660254038 * this.r);
    }

    public getCenterY(size: number = 1) {
        return size * 1.5 * this.r;
    }

    public getNeighborPosition(direction: HexDirection) {
        return HexPosition.add(this, HexPosition.getDirectionVector(direction));
    }

    public toString() {
        return '(' + this.q + ',' + this.r + ')';
    }

    public getPathString(size: number = 1) {
        const s = parseFloat(size.toString());
        const xOff = s * 0.8660254038;
        const halfSize = s * 0.5;
        return (+xOff) + ',' + (-halfSize) + ' '
            + (+xOff) + ',' + (+halfSize) + ' '
            + (0    ) + ',' + (+s       ) + ' '
            + (-xOff) + ',' + (+halfSize) + ' '
            + (-xOff) + ',' + (-halfSize) + ' '
            + (0    ) + ',' + (-s       );
    }
}
