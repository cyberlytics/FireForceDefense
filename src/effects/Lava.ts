import Effect from '../model/Effect';
import type LevelMap from '../model/LevelMap';
import type HexPosition from '../model/HexPosition';
import Fire from '../model/Fire';

export default class Lava extends Effect {
    readonly description = 'Lava: Description';
    readonly id = 'Lava';
    readonly name = 'Lava';
    readonly duration: number = 5;

    applyEffect(map: LevelMap, pos: HexPosition): void {
        map.getCellsAround(pos, 3).forEach(cell => {
            cell.fireIntensity = Fire.modify(cell.fireIntensity, 12);
        });
    }

}
