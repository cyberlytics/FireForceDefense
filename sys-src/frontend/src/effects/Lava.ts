import Effect from '../model/Effect';
import type LevelMap from '../model/LevelMap';
import type HexPosition from '../model/HexPosition';
import Fire from '../model/Fire';

export default class Lava extends Effect {
    readonly description = 'Lava: Description';
    readonly id = 'Lava';
    readonly name = 'Lava';
    readonly duration = 5;
    readonly range = 2;

    applyEffect(map: LevelMap, pos: HexPosition): void {
        map.getCellsAround(pos, this.range).forEach((cell) => {
            cell.fireIntensity = Fire.modify(cell.fireIntensity, 12);
        });
    }
}
