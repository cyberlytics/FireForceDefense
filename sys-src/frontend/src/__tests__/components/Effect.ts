import { mount } from '@vue/test-utils';
import Effect from '@components/effect.vue';
import EffectExecution from '@model/EffectExecution';
import Lava from '@effects/Lava';
import HexPosition from '@model/HexPosition';

describe('Effect.vue', () => {
    test('effect matches snapshot', () => {
        const wrapper = mount(Effect, {
            propsData: {
                effectExecution: new EffectExecution(new Lava(), new HexPosition(0, 0)),
                size: 60,
            },
        });
        expect(wrapper.element).toMatchSnapshot();
    });
});
