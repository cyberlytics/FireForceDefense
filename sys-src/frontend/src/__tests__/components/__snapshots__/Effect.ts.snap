// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Effect.vue effect matches snapshot 1`] = `
<g
  style="pointer-events: none;"
  transform="translate(0,0)"
>
  <!---->
   
  <circle
    cx="0"
    cy="0"
    fill="url(#effect-Lava)"
    r="216.33307652783935"
  />
   
  <!---->
</g>
`;
